### ISSUE TITLE
async/await triggers ScErrThreadNotAttached

### ISSUE THREAD

**Issue by mtornwall** [Starcounter/Starcounter#2796]
The .NET async/await pattern mixes with Starcounter code. The following crashes the application with `ScErrThreadNotAttached`:

    [Database]
    public class Foo {}

    // ... somewhere in a transaction scope ...
    await Task.Delay(1000);
    new Foo(); // exception

While my understanding of Starcounter internals and the .NET async model is quite limited, it would appear that the Starcounter threads don't have a [SynchronizationContext](https://msdn.microsoft.com/en-us/library/system.threading.synchronizationcontext(v=vs.110).aspx) attached. The continuation of the asynchronous task (the part that instantiates a new `Foo`) is then dispatched to the default `SynchronizationContext`, which schedules it to run on the .NET thread pool. I guess the .NET pool threads are not attached to Starcounter schedulers?

I realize that mixing asynchronous operations with transactions is nontrivial, so I'm not sure as to whether I should file this as a bug or a feature request.

**Comment by dan31** [comment]
Hi! That feature is no.1 feature already addressed in new version of Starcounter (so-called 'dnx'). Since you feel to be quite in that topic, you'll find a pleasure to talk more about this with Johan L. at the office this week :) He can tell you how it works in upcoming version.

Sent from Outlook<http://taps.io/outlookmobile>




On Mon, Jul 6, 2015 at 12:32 AM -0700, "Martin Törnwall" <notifications@github.com<mailto:notifications@github.com>> wrote:


The .NET async/await pattern mixes with Starcounter code. The following crashes the application with ScErrThreadNotAttached:

[Database]
public class Foo {}

// ... somewhere in a transaction scope ...
await Task.Delay(1000);
new Foo(); // exception


While my understanding of Starcounter internals and the .NET async model is quite limited, it would appear that the Starcounter threads don't have a SynchronizationContext<https://msdn.microsoft.com/en-us/library/system.threading.synchronizationcontext(v=vs.110).aspx> attached. The continuation of the asynchronous task (the part that instantiates a new Foo) is then dispatched to the default SynchronizationContext, which schedules it to run on the .NET thread pool. I guess the .NET pool threads are not attached to Starcounter schedulers?

I realize that mixing asynchronous operations with transactions is nontrivial, so I'm not sure as to whether I should file this as a bug or a feature request.

-

**Comment by mtornwall** [comment]
Thanks Dan! Is this new version available somewhere, or should I work around this problem for the time being?

**Comment by dan31** [comment]
This version is in early stage (and we'll keep you posted on its availability), so you have to workaround in 2.0. If you post here what exactly a task do you want to solve with async, @chrhol can guess for you the right pattern in 2.0.

Sent from Outlook<http://taps.io/outlookmobile>




On Mon, Jul 6, 2015 at 1:22 AM -0700, "Martin Törnwall" <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks Dan! Is this new version available somewhere, or should I work around this problem for the time being?

-

**Comment by mtornwall** [comment]
Basically I want to do something like this:

    async void Handle(Input.Foo input)
    {
        var result = await DoSomeWebRequests();
        this.Data.Foo = result.Foo;
        this.Data.Bar = result.Bar;
    }

I've tried to hack around this using an asynchronous `Http.GET`:

    void Handle(Input.Foo input)
    {
        Http.GET("http://some.site/some/path", null, null, (response, _) => {
            // Transaction captured in closure... but is this safe!?
            this.Transaction.Scope(() => {
                var result = UnpackResponse(response);
                this.Data.Foo = result.Foo;
                this.Data.Bar = result.Bar;
            });
        });
    }

But I have two issues with this:

1. The changes to `this.Data` don't seem to propagate to the client. I haven't had time to look deeply into this so I may very well be doing something wrong on my end.
2. It seems like I may be abusing the fact that the closure captured the transaction. :-)

**Comment by mtornwall** [comment]
@Starcounter-Jack

**Comment by chrhol** [comment]
> The changes to this.Data don't seem to propagate to the client. I haven't had time to look deeply into this so I may very well be doing something wrong on my end.

I guess this might be due to the delegate for handling the response is executed after the request that calls the input-handler is finished. 

> It seems like I may be abusing the fact that the closure captured the transaction. :-) I'm probably doing a bit too much shooting from the hip here?

This looks ok to me if you manually need to switch to a transaction that is already attached to the viewmodel. 

One thing that came to mind is that I'm not sure if the response delegate is always executed on the same scheduler as called from. If not, then using a transaction created on a different scheduler will fail.

Isn't it easier in this case to not have asynchronous calls to external requests but rather to wait for the response and update the viewmodel directly? I.e. using `Response response = Http.GET("http://127.0.0.1/fooapp/internal/work?args=...");` instead.

```
...
Response resp = Http.GET("http://127.0.0.1/fooapp/internal/work?args=...");
var result = UnpackResponse(response);
this.Data.Foo = result.Foo;
this.Data.Bar = result.Bar;
...
```

**Comment by mtornwall** [comment]
> Isn't it easier in this case to not have asynchronous calls to external requests but rather to wait for the response and update the viewmodel directly?

Running the requests synchronously is not an option for me as they can take on the order of seconds to tens of seconds. A similar problem occurs in the initial construction of the viewmodel. GET:ing synchronously there could potentially have the client waiting on a response for a minute (!) or more.

Is there any way of getting at the Starcounter scheduler so that I can manually submit tasks to run?

**Comment by chrhol** [comment]
Sorry for the late response. For some reason I do not get an email notification for updates in this thread. 

This is an interesting case I think, and maybe not that uncommon. I see two things that needs to be solved:

- Handling the response from the external request inside a scheduler AND in the context of the correct session and viewmodel. We have an API for scheduling tasks that runs inside a Starcounter thread, but there you cannot access the session/viewmodel. Only databasedata is accessible, and since the viewmodel is not available you cannot switch to the correct transaction.
There is also API for enumerating all sessions and get a callback to a delegate where the viewmodel is available, however I see no way to specify one specific session. Maybe @alemoi knows more here.

- Getting the changes in the viewmodel to the client.
Since the external request is async the original response to the client will already have been sent when the delegate is executed, which means that even if the delegate is executed on a correct thread with the session/viewmodel accessible, the only options are to either push changes if websockets are used, or the changes will be sent the next time the client sends an request.

**Comment by dan31** [comment]
@chrhol, for some reason, I don't get notifications from SOME of Github threads either (and no regularity really), but I thought it is only my glitch. Since it isn't, that is a big warning sign. Maybe that's the consequence of recent months github ddos-es.

**Comment by alemoi** [comment]
About delayed responses you can read here: http://starcounter.io/guides/network/handling-responses/
"Delayed or explicitly handled responses".
So its supported.
Regarding running tasks with specific session we can add an API on Session. So far you can do the following (inside `RunAsync` delegate):

```c#
try {
	// Setting new current session.
	session.StartUsing();
	
	// Running user delegate with session.
	action(...);
} finally {
	session.StopUsing();
}
```

**Comment by mtornwall** [comment]
Thanks for the input @chrhol and @alemoi!

I tried the solution @alemoi proposed above, but it didn't make any visible difference - i.e. still no change propagation to the client. I suppose this is explained by @chrhol's earlier remark:

> Since the external request is async the original response to the client will already have been sent when the delegate is executed, which means that even if the delegate is executed on a correct thread with the session/viewmodel accessible, the only options are to either push changes if websockets are used, or the changes will be sent the next time the client sends an request.

I can easily extend my Puppet implementation to regularly exchange "empty" patches with the server. Would doing so cause the server to piggyback changes onto the patch response?

**Comment by chrhol** [comment]
> I can easily extend my Puppet implementation to regularly exchange "empty" patches with the server. Would doing so cause the server to piggyback changes onto the patch response?

Yes, the response will include all changes that have not yet been sent.

**Comment by mtornwall** [comment]
For the record, the solution I ended up using is this:

1. The asynchronous operation is initiated, at the client's request, by triggering an internal handler using an asynchronous `Http.POST`.
2. When the asynchronous POST returns, the result is applied to the viewmodel by attaching the session as suggested by @alemoi and entering a transaction scope.
3. In order to refresh the client, I have set it up to send regular heartbeats (really just empty patches).

This solution is satisfactory, if a bit complicated. :-) If a fix is planned for dnx, this issue can be closed as far as I'm concerned.
